modal/_utils/blob_utils.py

```diff
@@ -366,6 +377,113 @@ def source():
         mode,
     )
 
+_FileUploadSource2 = Callable[[], ContextManager[BinaryIO]]
+
+@dataclasses.dataclass
+class FileUploadSpec2:
+    source: _FileUploadSource2
+    source_description: Union[str, Path]
+
+    path: str
+    # Raw (unencoded 32 byte) SHA256 sum per 8MiB file block
+    blocks_sha256: list[bytes]
+    mode: int  # file permission bits (last 12 bits of st_mode)
+    size: int
+
+
+    @staticmethod
+    async def from_path(
+        filename: Path,
+        mount_filename: PurePosixPath,
+        mode: Optional[int] = None,
+    ) -> "FileUploadSpec2":
+        # Python appears to give files 0o666 bits on Windows (equal for user, group, and global),
+        # so we mask those out to 0o755 for compatibility with POSIX-based permissions.
+        mode = mode or os.stat(filename).st_mode & (0o7777 if platform.system() != "Windows" else 0o7755)
+
+        def source():
+            return open(filename, "rb")
+
+        return await FileUploadSpec2._create(
+            source,
+            filename,
+            mount_filename,
+            mode,
+        )
+
+
+    @staticmethod
+    async def from_fileobj(
+        source_fp: BinaryIO,
+        mount_filename: PurePosixPath,
+        mode: int
+    ) -> "FileUploadSpec2":
+        def source():
+            new_fd = os.dup(source_fp.fileno())
```

Comment:
Even though the type system doesn't allow it, do we have some users passing in non-file-backed objects here? How to best cater for that?

Reply:
Yeah I think we need to support passing `io.BytesIO` here. We support that today.

Python is weird - `BytesIO` objects have the `fileno` function but it blows up at runtime:
```
In [18]: x = io.BytesIO(b"foo")

In [19]: "fileno" in dir(x)
Out[19]: True

In [20]: x.fileno()
---------------------------------------------------------------------------
UnsupportedOperation                      Traceback (most recent call last)
Cell In[20], line 1
----> 1 x.fileno()

UnsupportedOperation: fileno
```

Reply:
I want to tweak the type signature for these functions to indicate that we accept more types of file-like objects, not only `BinaryIO`.

It made me suspicious as to why the tests are passing that test for this functionality:

```
test/volume_test.py::test_volume_batch_upload[2] PASSED
```

Something is broken with the test setup, I think.

Reply:
Fixed the test setup now, PTAL